File: Data Structures & Algorithms/merge-intervals/submission-2.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        """
            [1,3], [1,5], [6,7]

            [1]
        """

        intervals = sorted(intervals, key =  lambda a: (a[0],a[1]))
        result = [intervals[0]]

        for i in range(1, len(intervals)):
            if result[-1][1] >= intervals[i][0]:
                result[-1][1] = max(result[-1][1], intervals[i][1])
            
            else:
                result.append(intervals[i])
        
        return result

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        /*
            k = 2
            
            X - 0
            Y - 2
            
            X Y Y Y X
                    |
                |


            0 1 2 3 4
            A A A B A B B
                |
                        |

            A - 4  
            B - 2
        */
        int n = s.size();
        int start = 0, curr = 0 ,maxFreq = 0, result = 0;
        unordered_map<int,int>mp;
        while(curr < n){
            mp[s[curr]]++;
            maxFreq = max(maxFreq,mp[s[curr]]);
            while((curr-start+1) - maxFreq > k){
                mp[s[start]]--;
                start++;
            }
            result = max(result,curr-start+1);
            curr++;
        }
        return result;
    }
};
